### DISCUSSION TITLE
Invoking User-Defined Functions In Postgres Schema

### DISCUSSION THREAD

**Discussion by sjbodzo** [SeaQL/sea-orm#1408] [Q&A] [answered]
Hey everyone, 

We are migrating to Sea ORM currently, and one of our challenges is that we have complex user defined functions. There are several layers to our platform, but at the SQL layer we rely heavily on invoking these functions. 

What is the recommended approach to call these functions and use them in Sea ORM queries against a Postgres schema? 

As an example: 

Consider I have a schema called `v5`, and a function within it that tells you if a player is cleared for the upcoming match. Let's call this function `player_is_cleared(p v5.player)` and say it returns a boolean. 

That function relies on complex logic that is hard to replicate by hand, and we really want to rely on the function instead of repeating its logic since it can change often. 

What I really need to do is something like: 
``` select * from v5.player p where player_is_cleared(p) ```

Is the best approach to use raw SQL in this case as shown [here](https://www.sea-ql.org/SeaORM/docs/basic-crud/raw-sql/#query-by-raw-sql)?

**Comment by sjbodzo** [comment, marked as answer]
Leaving this here as a comment to others. I was able to find a few approaches but this one seems best.

The best way I have found to do this is by defining an empty struct type that has `impl Iden` on it, because then you can do the following: 

```rust
#[derive(Iden)]
#[iden = "v5.player_is_cleared"]
struct PlayerIsCleared;

// OR
// impl Iden for PlayerIsCleared {
//     fn unquoted(&self, s: &mut dyn std::fmt::Write) {
//         write!(s, "v5.player_is_cleared").unwrap();
//     }
// }

Player::find()
.filter(
    Condition::all()
        .add(player::Column::Id.is_in(ids))
        .add(player_state::Column::Verified.eq(filter_args.verified))
        .add(
            Func::cust(PlayerIsCleared)
                .args([Expr::val("player")])
                .equals(filter_args.clear_status),
        )
```

**Reply by billy1624** [reply]
Hey @sjbodzo, that's correct! One minor nitpick.

```rs
#[derive(Iden)]
#[iden = "v5.player_is_cleared"]
struct PlayerIsCleared;
```

**Reply by tyt2y3** [reply]
Yes this is the design and usage in SeaQuery
